### DISCUSSION TITLE
Jobs not being non-deterministically processed upon starting

### DISCUSSION THREAD

**Discussion by xinluh** [taskforcesh/bullmq#1633] [Q&A]
Hello, we have a fairly standard worker setup, starting the work by instantiating the worker object:
```
new Worker('queue_name', job => { console.log('job started'); ... },  { concurrency: 10, connection: {...} })
console.log('worker started')
```

The issue we ran into is, once in a blue moon, the worker would just not process any jobs at all - no error message, nada just silent (i.e. in the code sample above, we will just not get the `job started` log at all when this happens). Restarting will always fix it, but it makes our deployment very unreliable and someone needs to monitor the queue processing correctly at every deploy. 

he mysterious thing is, we have 4 queues, and 1 queue never runs into this problem, the other 3 are problematic but not always at the same time (sometimes just 1 queue would fail). There is no difference on how the queues were set up, except that the non-problematic queue has only concurrency of 4 instead of 10 and we put significantly more (short) jobs put on it.

Do you have any pointer to where to start investigating this issue? I can confirm that the `new Worker(...)` line was ran because we do get the log line of `worker started`. We also subscribe to the worker error `worker.on('error', ...)` but we get no logs from that either.

**Comment by manast** [comment]
Well, it could be as simple as there are jobs that stay in active for ever, did you check that that is not the case? maybe your workers are never finishing and therefore no new jobs will be processed. Using a dashboard will help you in getting more insight on what is going on.

**Reply by xinluh** [reply]
@manast this is during a redeploy, i.e. the old workers are killed and new workers coming online. I can confirm that old workers were killed from seeing the log lines. Even if the old workers are not killed properly, that shouldn't prevent the new workers from grabbing jobs?

> Using a dashboard will help you in getting more insight on what is going on.

What specific dashboard / metric do you think would help here?

**Reply by manast** [reply]
You can use https://taskforce.sh/ you will see the status of the jobs and probably sort out why they are not being processed.

**Reply by xinluh** [reply]
@manast we do use taskforce, but metric is not illuminating. Either way, as I mentioned in the original post, we don't see any log statement from the job execution at all.

**Reply by manast** [reply]
But what can you tell about the status of the jobs when this happens? are there still active jobs?

**Reply by xinluh** [reply]
We'll can look next time this happens!

But I'm a little fuzzy on how this relates to the error here - even if there are active jobs, how would that prevent the new workers from picking up new jobs?

**Reply by manast** [reply]
If you have a max concurrency of for example 10, and you have 10 active jobs, then no new jobs will be picked up.

**Reply by xinluh** [reply]
@manast I see - that might makes sense (and seems to fit the pattern that the queue with the shortest job seems to not have problem). 

However, let's say this is what is happening: old workers getting killed (hence the job currently claimed by them will never finish) -> new worker comes online and never picks up any jobs because concurrency is maxed out, then shouldn't the jobs on the old workers be eventually be marked as stalled, allowing the jobs to be picked up by the new workers?

I would also add that we didn't have this problem before with bull, and upgrading to bullmq seems to coincide with this problem happening (though since this was a while ago the timeline is a little fuzzy).

**Reply by manast** [reply]
If the jobs get stalled due to workers being killed ungracefully (if you gracefully close your workers when shutting down the servers then it will wait for current jobs to finalize), then the stalled jobs will just be moved to wait from active, so they will not consume "concurrency" so to speak and they will be picked up normally. In any case, why don't you gather more information? you can easily see what is happening if you study which jobs are in the active state and why they are not progressing.

**Reply by xinluh** [reply]
> In any case, why don't you gather more information?

Definitely doing that, it's just that (as I mentioned above) this is happening non-deterministically, so we have to wait for this to happen again.

**Reply by xinluh** [reply]
@manast circling back! We checked Taskforce while this is happening. There is no active job in the queue when this problem happens. 
<img width="1185" alt="image" src="https://user-images.githubusercontent.com/9114601/214930709-64b379fc-5131-4c39-9d03-111459c19612.png">

Any thoughts on the next step to debug?

**Reply by manast** [reply]
Do you have any workers running and are they connected to Redis properly?

**Reply by manast** [reply]
Btw, it looks like you are using Bull and not BullMQ according to your screenshot.

**Reply by xinluh** [reply]
@manast  yes the workers are running and connected to Redis properly (we know because one of our 4 queues are processing jobs just fine, and we initiate the workers all the same place)

We are queue/processing jobs with bullmq but we upgraded from bull. Is there a missing upgrade step that I missed or it just taskforce not detecting the change?

**Reply by manast** [reply]
Have you possibly paused the queue recently? we had some issues regarding pause and Bull <-> BullMQ interoperability.

**Reply by manast** [reply]
We found an issue that maybe is the cause to your issue https://github.com/taskforcesh/bullmq/pull/1654
A release is coming as soon as the PR is green and approved.

**Reply by manast** [reply]
https://bullmq.io/news/022023/connection-issue/

**Reply by xinluh** [reply]
Thanks @manast we'll do the update and see if that solves the problem!

Re: bullmq - is there something we can do to fully upgrade from bull with no or minimal interruption to the queue?

**Reply by manast** [reply]
Although BullMQ is mostly backward compatible, the safest would be to use new queues if you plan to move to BullMQ, as we do not guarantee full backwards compatibility.

**Reply by xinluh** [reply]
@manast bad news :/ We upgraded to the version with the fix: it did not solve the non-deterministic not processing job problem. (It did improved the delay between job queued and processed though: I noticed that the latency got worse by a few seconds when upgrading from bull -> bullmq; with the new version, the latency is back to a low level like in bull).

Do you have suggestions on what/how to dig in next to figure out what is going on?

**Reply by manast** [reply]
No sorry. The best is if you provide a simple test case that reproduces the issue. Something must be special in your use case as this issue does not seem to happen to anybody else 🤔

**Reply by manast** [reply]
Have you tried a dashboard to get better visibility on what is going on? You can use taskforce.sh for free, it may help you to find the issue, if not let me know and we will help you.

**Reply by xinluh** [reply]
@manast as I mentioned above, we do use taskforce.sh (and posted a screenshot above). But it doesn't quite help us with useful information in this case. I'm going to try to fix the not graceful worker shutdown problem next (it is problematic for other reasons though unclear to me that it could cause the non-deterministic not processing job problem).

And I can try to fully upgrade to bullmq queues just to eliminate one more potential cause - do you have suggestion on how to do this in a non-disruptive way? Sounds like we need to somehow spin up new queues to force the upgrade (even though we are using bullmq library)? We have jobs getting scheduled at every second and it's not an option to stop those jobs from being scheduled.

**Reply by manast** [reply]
From which version do you need to upgrade?

**Reply by xinluh** [reply]
We were on bull 4.9.0, then we upgraded to using bullmq (it was 2.1.0 when we upgraded, now we are on 3.6.6), pointing to the same redis and queue names. I didn't realize that that would not upgrade the underlying queue to bullmq queues (actually didn't even realize that the underlying queues are different between bull and bullmq).

**Reply by manast** [reply]
They are slightly different but besides the operation of pausing, old Bull queues will work with BullMQ.

**Reply by manast** [reply]
In any case, check the changelog as there are some breaking changes that you should be aware of even if they wont affect you if you are not using rate-limit group keys: https://docs.bullmq.io/changelog#3.0.0-2022-10-25
